### PR TITLE
feat: token management

### PR DESCRIPTION
**Pull request by wansir**
**What type of PR is this?**

/kind feature
/aera iam

**What this PR does / why we need it**:

1. add token management logic.
2. support sliding session, auto-Increment expiration time when token is used.

### CODE REVIEW

**Comment by zryfish** [comment]
Do we have to use redis to manage user sessions ? we are trying to reduce component dependency in our system.

**Comment by wansir** [comment]
> Do we have to use redis to manage user sessions ? we are trying to reduce component dependency in our system.

We still need to share session data through redis unless ks-account uses only one replica.

**Comment by zryfish** [comment]
> > Do we have to use redis to manage user sessions ? we are trying to reduce component dependency in our system.
> 
> We still need to share session data through redis unless ks-account uses only one replica.

Can we make redis optional ? If redis provided, then use redis as session backend, otherwise just use memory.  It would be better if can eliminate usage of redis in allinone mode

**Comment by wansir** [comment]
We also need to share the sessions between ks-account and ks-apigateway via redis and increase session expiration time based on API access.

**Comment by wansir** [comment]
> > > Do we have to use redis to manage user sessions ? we are trying to reduce component dependency in our system.
> > 
> > 
> > We still need to share session data through redis unless ks-account uses only one replica.
> 
> Can we make redis optional ? If redis provided, then use redis as session backend, otherwise just use memory. It would be better if we can eliminate usage of redis in allinone mode

We also need to share the sessions between ks-account and ks-apigateway via redis and increase session expiration time based on API access.

**Comment by zryfish** [comment]
> > > > Do we have to use redis to manage user sessions ? we are trying to reduce component dependency in our system.
> > > 
> > > 
> > > We still need to share session data through redis unless ks-account uses only one replica.
> > 
> > 
> > Can we make redis optional ? If redis provided, then use redis as session backend, otherwise just use memory. It would be better if we can eliminate usage of redis in allinone mode
> 
> We also need to share the sessions between ks-account and ks-apigateway via redis and increase session expiration time based on API access.

So this is the part I can't understand. We do session and token management in ks-console, ks-account, ks-apigateway. That seems not right. We should do that in one place, such as ks-apigateway, ks-account is only responsible for authentication.

File: pkg/simple/client/redis/options.go [hunk 1]
```diff
@@ -1,67 +1,47 @@
...
-	"kubesphere.io/kubesphere/pkg/utils/net"
 	"kubesphere.io/kubesphere/pkg/utils/reflectutils"
 )
 
 type RedisOptions struct {
-	Host     string
-	Port     int
-	Password string
-	DB       int
+	RedisURL string
```
[the remark this answers is not in the document]
Comment (zryfish) [reply pkg/simple/client/redis/options.go]: yeah, but it's more complicated for user to configure

File: pkg/simple/client/redis/options.go [hunk 2]
```diff
@@ -1,67 +1,47 @@
...
 		reflectutils.Override(options, r)
 	}
 }
 
 // AddFlags add option flags to command line flags,
 // if redis-host left empty, the following options will be ignored.
 func (r *RedisOptions) AddFlags(fs *pflag.FlagSet) {
-	fs.StringVar(&r.Host, "redis-host", r.Host, ""+
+	fs.StringVar(&r.RedisURL, "redis-url", "", ""+
 		"Redis service host address. If left blank, means redis is unnecessary, "+
```
Comment (zryfish) [inline pkg/simple/client/redis/options.go]: need an example redis-url in description

**Pull request merged**